segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java

```diff
@@ -106,19 +163,134 @@ public void stop() {
+        long start = MathUtils.now();
+        String[] split = address.split(":");
+        String host = split[0];
+        int port = Integer.parseInt(split[1]);
+
+        while (true) {
+            try {
+                SSLContext context = SSLContext.getInstance("TLS");
+                TrustManagerFactory trustManager = getTrustManager(trustStore, trustStorePasswordPath);
+                KeyManagerFactory keyFactory = getKeyManager(keyStore, keyStorePasswdPath);
+
+                context.init(keyFactory.getKeyManagers(), trustManager.getTrustManagers(), null);
+                SSLSocketFactory factory = context.getSocketFactory();
+
+                Socket sock = context.getSocketFactory().createSocket(new Socket(host, port), host, port, true);
+                BufferedReader reader = null;
+                try {
+                    OutputStream outstream = sock.getOutputStream();
+                    outstream.write("stat".getBytes());
+                    outstream.flush();
+
+                    reader =
+                            new BufferedReader(
+                                    new InputStreamReader(sock.getInputStream()));
+                    String line = reader.readLine();
+                    if (line != null && line.startsWith("Zookeeper version:")) {
+                        log.info("Server UP");
+                        return true;
+                    }
+                } catch (Exception e) {
+                  e.printStackTrace();
+                } finally {
+                    sock.close();
+                    if (reader != null) {
+                        reader.close();
+                    }
+                }
+            } catch (IOException | CertificateException | NoSuchAlgorithmException | KeyStoreException
+                    | KeyManagementException | UnrecoverableKeyException e) {
+                // ignore as this is expected
+                log.info("server " + address + " not up " + e);
+            }
+
+            if (MathUtils.now() > start + timeout) {
+                break;
+            }
+            try {
+                Thread.sleep(250);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return false;
     }
 
+    private static TrustManagerFactory getTrustManager(String trustStore, String trustStorePasswordPath)
+            throws IOException, CertificateException, NoSuchAlgorithmException, KeyStoreException {
+        try (FileInputStream myKeys = new FileInputStream(trustStore)) {
+
+            // Do the same with your trust store this time
+            // Adapt how you load the keystore to your needs
+            KeyStore myTrustStore = KeyStore.getInstance("JKS");
+            myTrustStore.load(myKeys, loadPasswdFromFile(trustStorePasswordPath).toCharArray());
+
+            myKeys.close();
+
+            TrustManagerFactory tmf = TrustManagerFactory
```

Comment:
one line

Reply:
Also move these 2-3 lines outside of the `try` block. This way you don't need the `myKeys.close()` call above.